### PR TITLE
fix: create volume sanity test failures

### LINKED ISSUE #27: fix sanity test failures
**What happened**:

```
Summarizing 5 Failures:

[Fail] Controller Service CreateVolume [It] should not fail when requesting to create a volume with already existing name and same capacity. 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:371

[Fail] Controller Service CreateVolume [It] should fail when requesting to create a volume with already existing name and different capacity. 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:442

[Fail] Controller Service ValidateVolumeCapabilities [It] should fail when the requested volume does not exist 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:705

[Fail] Node Service [It] should work 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/node.go:464

Ran 26 of 57 Specs in 77.639 seconds
FAIL! -- 21 Passed | 5 Failed | 0 Pending | 31 Skipped
```

following issue has been fixed:
```
[Fail] Controller Service DeleteVolume [It] should succeed when an invalid volume id is used 
/home/priyanshu/work/src/github.com/csi-driver/azurefile-csi-driver/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:543
```
**What you expected to happen**:

**How to reproduce it**:
run sanity test: https://github.com/kubernetes-sigs/azurefile-csi-driver/tree/master/test/sanity

**Anything else we need to know?**:

**Environment**:
- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by ZeroMagic**
**What type of PR is this?**
/kind bug
/kind failing-test

**What this PR does / why we need it**:

- set file share name correctly
- save the created volume so that we can know whether the volume exists or not

**Which issue(s) this PR fixes**:

Fixes #27 

**Special notes for your reviewer**:

**Release note**:
```
fix: create volume sanity test failures
```

### CODE REVIEW

File: pkg/azurefile/controllerserver.go [hunk 1]
```diff
@@ -94,6 +91,21 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 		return nil, fmt.Errorf("failed to create file share(%s) on account(%s) type(%s) rg(%s) location(%s) size(%d), error: %v", fileShareName, account, sku, resourceGroup, location, requestGiB, err)
 	}
 	volumeID := fmt.Sprintf(volumeIDTemplate, resourceGroup, retAccount, fileShareName)
+	capacityBytes := req.GetCapacityRange().GetRequiredBytes()
+
+	if vol, ok := d.volumes[volumeID]; ok {
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: The check should happen before `d.cloud.CreateFileShare...`

File: pkg/azurefile/azurefile.go [hunk 2]
```diff
@@ -54,6 +61,7 @@ func NewDriver(nodeID string) *Driver {
 	driver.Name = DriverName
 	driver.Version = driverVersion
 	driver.NodeID = nodeID
+	driver.volumes = make(map[string]*volume)
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go]: store all volumes in memory is not a good solution, what happen if this driver crash or reschedule to other node

File: pkg/azurefile/controllerserver.go [hunk 3]
```diff
@@ -82,10 +80,9 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 		accountKind = string(storage.FileStorage)
 	}
 
-	if fileShareName == "" {
-		// File share name has a length limit of 63, and it cannot contain two consecutive '-'s.
-		fileShareName = util.GenerateVolumeName("pvc-file", uuid.NewUUID().String(), 63)
-		fileShareName = strings.Replace(fileShareName, "--", "-", -1)
+	fileShareName, err := setFileShareName(fileShareName, name)
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: if fileShareName == "" {
    fileShareName = getValidFileShareName(name)
}

File: pkg/azurefile/controllerserver.go [hunk 4]
```diff
@@ -213,3 +237,35 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
 func (d *Driver) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "ControllerExpandVolume is not yet implemented")
 }
+
+// File share names can contain only lowercase letters, numbers, and hyphens,
+// and must begin and end with a letter or a number,
+// and must be from 3 through 63 characters long.
+// The name cannot contain two consecutive hyphens.
+func setFileShareName(fileShareName, reqName string) (string, error) {
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: rename to `getValidFileShareName` and should not return error, we do the filtering, and if original vol name is still invalid, generte a old UUID style share name, and also write a unit test for this func

File: pkg/azurefile/controllerserver.go [hunk 5]
```diff
@@ -213,3 +237,35 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
 func (d *Driver) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "ControllerExpandVolume is not yet implemented")
 }
+
+// File share names can contain only lowercase letters, numbers, and hyphens,
```
Comment (andyzhangx) [inline pkg/azurefile/controllerserver.go]: it's better add a doc link about the file share name rule

**Comment by andyzhangx** [comment]
@ZeroMagic about how to get existing file share API, you could refer to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/azure/azure_file.go#L82, add a new func like `(d *Driver) CheckFileShare(accountName, accountKey, name string, sizeGiB int) error`

**Comment by ZeroMagic** [comment]
Done.

File: pkg/azurefile/controllerserver.go [hunk 6]
```diff
@@ -213,3 +233,72 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
 func (d *Driver) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "ControllerExpandVolume is not yet implemented")
 }
+
+// CheckFileShare check if the volume exists
+func (d *Driver) CheckFileShare(accountName, resourceGroup, name string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azurefile/controllerserver.go]: and do you also want to validate the `capacityBytes` in this func?

[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: Currently, I have no idea how to get capacityBytes of the existing volume. So in this pr, we can skip the problem about capacityBytes.

File: pkg/azurefile/controllerserver.go [hunk 7]
```diff
@@ -33,6 +34,11 @@ import (
 	"google.golang.org/grpc/status"
 )
 
+var (
+	fileShareNameMinLength = 3
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: done.

File: pkg/azurefile/controllerserver.go [hunk 8]
```diff
@@ -83,9 +91,21 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
...
-		fileShareName = util.GenerateVolumeName("pvc-file", uuid.NewUUID().String(), 63)
-		fileShareName = strings.Replace(fileShareName, "--", "-", -1)
+		fileShareName = getValidFileShareName(name)
+	}
+
+	exist, err := d.CheckFileShare(account, resourceGroup, fileShareName)
+	if err != nil {
+		return nil, err
+	}
+	if exist {
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: done.

**Comment by andyzhangx** [comment]
```
hack/verify-update.sh
Verifying dependency update
Some files got changed after dependencies update
 vendor/modules.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
Makefile:58: recipe for target 'verify-modules' failed
make: *** [verify-modules] Error 1
```

File: pkg/azurefile/azurefile_test.go [hunk 9]
```diff
@@ -214,3 +214,77 @@ func TestGetStorageAccount(t *testing.T) {
...
 	}
 }
+
+func TestGetValidFileShareName(t *testing.T) {
+	tests := []struct {
+		options  string
+		expected string
+	}{
+		{
+			options:  "aqz",
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile_test.go]: add empty string condition

File: pkg/azurefile/azurefile_test.go [hunk 10]
```diff
@@ -214,3 +214,77 @@ func TestGetStorageAccount(t *testing.T) {
 		}
 	}
 }
+
+func TestGetValidFileShareName(t *testing.T) {
+	tests := []struct {
+		options  string
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile_test.go]: `options` --> `volumeName`

File: pkg/azurefile/azurefile_test.go [hunk 11]
```diff
@@ -214,3 +214,77 @@ func TestGetStorageAccount(t *testing.T) {
...
+		result := getValidFileShareName(test.options)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("input: %q, getValidFileShareName result: %q, expected: %q", test.options, result, test.expected)
+		}
+	}
+}
+
+func TestCheckShareNameBeginAndEnd(t *testing.T) {
+	tests := []struct {
+		options  string
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile_test.go]: `options` --> `fileShareName`

File: pkg/azurefile/azurefile.go [hunk 12]
```diff
@@ -101,6 +108,40 @@ func (d *Driver) Run(endpoint string) {
...
+// CheckFileShareExists check if the volume exists
+func (d *Driver) CheckFileShareExists(accountName, resourceGroup, name string) (bool, error) {
+	// find the access key with this account
+	accountKey, err := d.cloud.GetStorageAccesskey(accountName, resourceGroup)
+	if err != nil {
+		// Because the account does not exist in the resource group, the file share could not exist.
+		// In the next step, we will create a temp account to create file share.
+		// So here we return nil, and warn the user.
+		klog.Warningf("could not get storage key for storage account %s: %v", accountName, err)
+		return false, nil
```
[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azurefile/azurefile.go]: I was wrong, keep the code and comments.

File: pkg/azurefile/azurefile.go [hunk 13]
```diff
@@ -101,6 +108,40 @@ func (d *Driver) Run(endpoint string) {
...
+	}
+
+	fileClient, err := d.getFileSvcClient(accountName, accountKey)
+	if err != nil {
+		return false, fmt.Errorf("error creating azure client: %v", err)
+	}
+	if exists, err := fileClient.GetShareReference(name).Exists(); err != nil {
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go]: what about 
```
return fileClient.GetShareReference(name).Exists()
```

File: pkg/azurefile/azurefile.go [hunk 14]
```diff
@@ -180,3 +221,36 @@ func getStorageAccount(secrets map[string]string) (string, string, error) {
...
 }
+
+// File share names can contain only lowercase letters, numbers, and hyphens,
+// and must begin and end with a letter or a number,
+// and must be from 3 through 63 characters long.
+// The name cannot contain two consecutive hyphens.
+//
+// See https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-shares--directories--files--and-metadata#share-names
+func getValidFileShareName(volumeName string) string {
+	fileShareName := strings.Replace(volumeName, "--", "-", -1)
```
Comment (andyzhangx) [inline pkg/azurefile/azurefile.go]: remove this and move following code into the end of this function:
```
fileShareName = strings.Replace(fileShareName, "--", "-", -1)
```
so we only need one Replace action.

**Comment by andyzhangx** [comment]
/hold
There are now 9 sanity test failures now, original it's 7 failures.

File: pkg/azurefile/azurefile.go [hunk 15]
```diff
@@ -101,6 +108,40 @@ func (d *Driver) Run(endpoint string) {
 	s.Wait()
 }
 
+// CheckFileShareExists check if the volume exists
+func (d *Driver) CheckFileShareExists(accountName, resourceGroup, name string) (bool, error) {
+	// find the access key with this account
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/azurefile.go]: Do we need to convert this function to private ?

[the remark this answers is not in the document]
Comment (andyzhangx) [reply pkg/azurefile/azurefile.go]: > Do we need to convert this function to private ?

make it private is better

File: pkg/azurefile/azurefile_test.go (hunk 9 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile_test.go]: If the string is empty, we will generate a uuid string. Then we can not know the expected string.

File: pkg/azurefile/azurefile.go (hunk 13 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile.go]: That's a good idea

File: pkg/azurefile/azurefile.go (hunk 14 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile.go]: If in this way, in the warning logging, user maybe get a wrong filesShare name.

Comment (andyzhangx) [reply pkg/azurefile/azurefile.go]: > If in this way, in the warning logging, user maybe get a wrong filesShare name.

I think that's ok, it's only a log

File: pkg/azurefile/controllerserver.go [hunk 16]
```diff
@@ -83,9 +85,19 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
...
-		// File share name has a length limit of 63, and it cannot contain two consecutive '-'s.
-		fileShareName = util.GenerateVolumeName("pvc-file", uuid.NewUUID().String(), 63)
-		fileShareName = strings.Replace(fileShareName, "--", "-", -1)
+		fileShareName = getValidFileShareName(name)
+	}
+
+	if exists, err := d.CheckFileShareExists(account, resourceGroup, fileShareName); err != nil {
+		return nil, err
+	} else if exists {
+		return &csi.CreateVolumeResponse{
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/controllerserver.go]: done.

File: pkg/azurefile/azurefile.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azurefile/azurefile.go]: done

File: pkg/azurefile/azurefile_test.go (hunk 10 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile_test.go]: done

File: pkg/azurefile/azurefile_test.go (hunk 11 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile_test.go]: done

File: pkg/azurefile/azurefile.go (hunk 14 above)
Comment (ZeroMagic) [reply pkg/azurefile/azurefile.go]: done

**Pull request merged**